File: dotnet/src/Grafana.Sigil/GenerationValidator.cs
namespace Grafana.Sigil;

public static class GenerationValidator
{
    public static void Validate(Generation generation)
    {
        if (generation.Mode is not null && generation.Mode is not (GenerationMode.Sync or GenerationMode.Stream))
        {
            throw new ArgumentException("generation.mode must be one of SYNC|STREAM");
        }

        if (string.IsNullOrWhiteSpace(generation.Model.Provider))
        {
            throw new ArgumentException("generation.model.provider is required");
        }

        if (string.IsNullOrWhiteSpace(generation.Model.Name))
        {
            throw new ArgumentException("generation.model.name is required");
        }

        for (var i = 0; i < generation.Input.Count; i++)
        {
            ValidateMessage("generation.input", i, generation.Input[i]);
        }

        for (var i = 0; i < generation.Output.Count; i++)
        {
            ValidateMessage("generation.output", i, generation.Output[i]);
        }

        for (var i = 0; i < generation.Tools.Count; i++)
        {
            if (string.IsNullOrWhiteSpace(generation.Tools[i].Name))
            {
                throw new ArgumentException($"generation.tools[{i}].name is required");
            }
        }

        for (var i = 0; i < generation.Artifacts.Count; i++)
        {
            ValidateArtifact(i, generation.Artifacts[i]);
        }
    }

    private static void ValidateMessage(string path, int messageIndex, Message message)
    {
        if (message.Role is not (MessageRole.User or MessageRole.Assistant or MessageRole.Tool))
        {
            throw new ArgumentException($"{path}[{messageIndex}].role must be one of user|assistant|tool");
        }

        if (message.Parts.Count == 0)
        {
            throw new ArgumentException($"{path}[{messageIndex}].parts must not be empty");
        }

        for (var i = 0; i < message.Parts.Count; i++)
        {
            ValidatePart(path, messageIndex, i, message.Role, message.Parts[i]);
        }
    }

    private static void ValidatePart(string path, int messageIndex, int partIndex, MessageRole role, Part part)
    {
        if (part.Kind is not (PartKind.Text or PartKind.Thinking or PartKind.ToolCall or PartKind.ToolResult))
        {
            throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}].kind is invalid");
        }

        var fieldCount = 0;
        if (!string.IsNullOrWhiteSpace(part.Text))
        {
            fieldCount++;
        }

        if (!string.IsNullOrWhiteSpace(part.Thinking))
        {
            fieldCount++;
        }

        if (part.ToolCall != null)
        {
            fieldCount++;
        }

        if (part.ToolResult != null)
        {
            fieldCount++;
        }

        if (fieldCount != 1)
        {
            throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}] must set exactly one payload field");
        }

        switch (part.Kind)
        {
            case PartKind.Text when string.IsNullOrWhiteSpace(part.Text):
                throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}].text is required");
            case PartKind.Thinking:
                if (role != MessageRole.Assistant)
                {
                    throw new ArgumentException(
                        $"{path}[{messageIndex}].parts[{partIndex}].thinking only allowed for assistant role"
                    );
                }

                if (string.IsNullOrWhiteSpace(part.Thinking))
                {
                    throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}].thinking is required");
                }

                break;
            case PartKind.ToolCall:
                if (role != MessageRole.Assistant)
                {
                    throw new ArgumentException(
                        $"{path}[{messageIndex}].parts[{partIndex}].tool_call only allowed for assistant role"
                    );
                }

                if (part.ToolCall == null || string.IsNullOrWhiteSpace(part.ToolCall.Name))
                {
                    throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}].tool_call.name is required");
                }

                break;
            case PartKind.ToolResult:
                if (role != MessageRole.Tool)
                {
                    throw new ArgumentException(
                        $"{path}[{messageIndex}].parts[{partIndex}].tool_result only allowed for tool role"
                    );
                }

                if (part.ToolResult == null)
                {
                    throw new ArgumentException($"{path}[{messageIndex}].parts[{partIndex}].tool_result is required");
                }

                break;
        }
    }

    private static void ValidateArtifact(int index, Artifact artifact)
    {
        if (artifact.Kind is not (
                ArtifactKind.Request
                or ArtifactKind.Response
                or ArtifactKind.Tools
                or ArtifactKind.ProviderEvent
            ))
        {
            throw new ArgumentException($"generation.artifacts[{index}].kind is invalid");
        }

        if (string.IsNullOrWhiteSpace(artifact.RecordId) && artifact.Payload.Length == 0)
        {
            throw new ArgumentException($"generation.artifacts[{index}] must provide payload or record_id");
        }
    }

    public static void ValidateEmbeddingStart(EmbeddingStart start)
    {
        var value = start ?? new EmbeddingStart();
        if (string.IsNullOrWhiteSpace(value.Model.Provider))
        {
            throw new ArgumentException("embedding.model.provider is required");
        }

        if (string.IsNullOrWhiteSpace(value.Model.Name))
        {
            throw new ArgumentException("embedding.model.name is required");
        }

        if (value.Dimensions.HasValue && value.Dimensions.Value <= 0)
        {
            throw new ArgumentException("embedding.dimensions must be > 0");
        }

        if (value.EncodingFormat.Length > 0 && string.IsNullOrWhiteSpace(value.EncodingFormat))
        {
            throw new ArgumentException("embedding.encoding_format must not be blank");
        }
    }

    public static void ValidateEmbeddingResult(EmbeddingResult result)
    {
        var value = result ?? new EmbeddingResult();
        if (value.InputCount < 0)
        {
            throw new ArgumentException("embedding.input_count must be >= 0");
        }

        if (value.InputTokens < 0)
        {
            throw new ArgumentException("embedding.input_tokens must be >= 0");
        }

        if (value.Dimensions.HasValue && value.Dimensions.Value <= 0)
        {
            throw new ArgumentException("embedding.dimensions must be > 0");
        }
    }
}

File: java/core/src/main/java/com/grafana/sigil/sdk/GenerationValidator.java
package com.grafana.sigil.sdk;

import java.util.List;

final class GenerationValidator {
    private GenerationValidator() {
    }

    static void validate(Generation generation) {
        boolean contentStripped = SigilClient.isContentStripped(generation);

        if (generation.getMode() != GenerationMode.SYNC && generation.getMode() != GenerationMode.STREAM) {
            throw new ValidationException("generation.mode must be one of SYNC|STREAM");
        }
        if (generation.getModel().getProvider().isBlank()) {
            throw new ValidationException("generation.model.provider is required");
        }
        if (generation.getModel().getName().isBlank()) {
            throw new ValidationException("generation.model.name is required");
        }

        validateMessages("generation.input", generation.getInput(), contentStripped);
        validateMessages("generation.output", generation.getOutput(), contentStripped);

        for (int i = 0; i < generation.getTools().size(); i++) {
            ToolDefinition tool = generation.getTools().get(i);
            if (tool == null || tool.getName().isBlank()) {
                throw new ValidationException("generation.tools[" + i + "].name is required");
            }
        }

        for (int i = 0; i < generation.getArtifacts().size(); i++) {
            Artifact artifact = generation.getArtifacts().get(i);
            if (artifact == null) {
                throw new ValidationException("generation.artifacts[" + i + "] is required");
            }
            if (artifact.getKind() == null) {
                throw new ValidationException("generation.artifacts[" + i + "].kind is invalid");
            }
            if (artifact.getRecordId().isBlank() && artifact.getPayload().length == 0) {
                throw new ValidationException("generation.artifacts[" + i + "] must provide payload or record_id");
            }
        }
    }

    private static void validateMessages(String path, List<Message> messages, boolean contentStripped) {
        for (int i = 0; i < messages.size(); i++) {
            Message message = messages.get(i);
            if (message == null || message.getRole() == null) {
                throw new ValidationException(path + "[" + i + "].role must be one of user|assistant|tool");
            }
            if (message.getParts().isEmpty()) {
                throw new ValidationException(path + "[" + i + "].parts must not be empty");
            }
            for (int j = 0; j < message.getParts().size(); j++) {
                validatePart(path, i, j, message.getRole(), message.getParts().get(j), contentStripped);
            }
        }
    }

    private static void validatePart(String path, int messageIndex, int partIndex, MessageRole role, MessagePart part, boolean contentStripped) {
        if (part == null || part.getKind() == null) {
            throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].kind is invalid");
        }

        boolean strippedTextOrThinking = contentStripped
                && (part.getKind() == MessagePartKind.TEXT || part.getKind() == MessagePartKind.THINKING);

        int payloadFieldCount = 0;
        if (!part.getText().isBlank()) {
            payloadFieldCount++;
        }
        if (!part.getThinking().isBlank()) {
            payloadFieldCount++;
        }
        if (part.getToolCall() != null) {
            payloadFieldCount++;
        }
        if (part.getToolResult() != null) {
            payloadFieldCount++;
        }
        if (payloadFieldCount != 1 && !strippedTextOrThinking) {
            throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "] must set exactly one payload field");
        }

        switch (part.getKind()) {
            case TEXT -> {
                if (part.getText().isBlank() && !contentStripped) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].text is required");
                }
            }
            case THINKING -> {
                if (role != MessageRole.ASSISTANT) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].thinking only allowed for assistant role");
                }
                if (part.getThinking().isBlank() && !contentStripped) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].thinking is required");
                }
            }
            case TOOL_CALL -> {
                if (role != MessageRole.ASSISTANT) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].tool_call only allowed for assistant role");
                }
                if (part.getToolCall() == null || part.getToolCall().getName().isBlank()) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].tool_call.name is required");
                }
            }
            case TOOL_RESULT -> {
                if (role != MessageRole.TOOL) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].tool_result only allowed for tool role");
                }
                if (part.getToolResult() == null) {
                    throw new ValidationException(path + "[" + messageIndex + "].parts[" + partIndex + "].tool_result is required");
                }
            }
        }
    }
}
